Mathlib/NumberTheory/Padics/AddChar.lean

```diff
@@ -0,0 +1,107 @@
+variable {p : ℕ} [Fact p.Prime]
+
+/-- If two continuous additive characters of `ℤ_[p]` agree at 1, they agree everwhere. -/
+lemma AddChar.eq_of_apply_one_eq {M : Type*} [Monoid M] [TopologicalSpace M] [T2Space M]
+    {κ₁ κ₂ : AddChar ℤ_[p] M} (hκ₁ : Continuous κ₁) (hκ₂ : Continuous κ₂) (h : κ₁ 1 = κ₂ 1) :
+    κ₁ = κ₂ := by
+  refine DFunLike.coe_injective <| PadicInt.denseRange_natCast.equalizer hκ₁ hκ₂ (funext fun n ↦ ?_)
+  simp only [Function.comp_apply, ← nsmul_one, h, AddChar.map_nsmul_eq_pow]
+
+variable {R : Type*} [NormedRing R] [Algebra ℤ_[p] R] [IsBoundedSMul ℤ_[p] R]
+  [IsUltrametricDist R]
+
+lemma AddChar.tendsto_apply_one_sub_pow {κ : AddChar ℤ_[p] R} (hκ : Continuous κ) :
+    Tendsto (fun n ↦ (κ 1 - 1) ^ n) atTop (𝓝 0) := by
+  refine (PadicInt.fwdDiff_tendsto_zero ⟨κ, hκ⟩).congr fun n ↦ ?_
+  simpa only [AddChar.map_zero_eq_one, mul_one] using fwdDiff_addChar_eq κ 0 1 n
+
+namespace PadicInt
+variable [CompleteSpace R]
+
+/-- The unique continuous additive character of `ℤ_[p]` mapping `1` to `1 + r`. -/
+noncomputable def addChar_of_value_at_one (r : R) (hr : Tendsto (r ^ ·) atTop (𝓝 0)) :
+    AddChar ℤ_[p] R where
+  toFun := mahlerSeries (r ^ ·)
+  map_zero_eq_one' := by
+    rw [← Nat.cast_zero, mahlerSeries_apply_nat hr le_rfl, zero_add, Finset.sum_range_one,
+      Nat.choose_self, pow_zero, one_smul]
+  map_add_eq_mul' a b := by
+    let F : C(ℤ_[p], R) := mahlerSeries (r ^ ·)
+    show F (a + b) = F a * F b
+    -- It is fiddly to show directly that `F (a + b) = F a * F b` for general `a, b`,
+    -- so we prove it for `a, b ∈ ℕ` directly, and then deduce it for all `a, b` by continuity.
+    have hF (n : ℕ) : F n = (r + 1) ^ n := by
+      rw [mahlerSeries_apply_nat hr le_rfl, (Commute.one_right _).add_pow]
+      refine Finset.sum_congr rfl fun i hi ↦ ?_
+      rw [one_pow, mul_one, nsmul_eq_mul, Nat.cast_comm]
+    refine congr_fun ((denseRange_natCast.prodMap denseRange_natCast).equalizer
+      ((map_continuous F).comp continuous_add)
+      (continuous_mul.comp (map_continuous <| F.prodMap F)) (funext fun ⟨m, n⟩ ↦ ?_)) (a, b)
+    simp [← Nat.cast_add, hF, ContinuousMap.prodMap_apply, pow_add]
+
+lemma continuous_addChar_of_value_at_one {r : R} (hr : Tendsto (r ^ ·) atTop (𝓝 0)) :
+    Continuous (addChar_of_value_at_one r hr : ℤ_[p] → R) :=
+  map_continuous (mahlerSeries (r ^ ·))
+
+lemma coe_addChar_of_value_at_one {r : R} (hr : Tendsto (r ^ ·) atTop (𝓝 0)) :
+    (addChar_of_value_at_one r hr : ℤ_[p] → R) = mahlerSeries (r ^ ·) :=
+  rfl
+
+lemma addChar_of_value_at_one_def {r : R} (hr : Tendsto (r ^ ·) atTop (𝓝 0)) :
+    addChar_of_value_at_one r hr (1 : ℤ_[p]) = 1 + r := by
+  show mahlerSeries (r ^ ·) ↑(1 : ℕ) = _
+  rw [mahlerSeries_apply_nat hr le_rfl, Finset.sum_range_succ, Finset.sum_range_one,
+    Nat.choose_zero_right, Nat.choose_self, one_smul, one_smul, pow_zero, pow_one]
+
+variable (p R) in
+/-- Equivalence between continuous additive characters `ℤ_[p] → R`, and `r ∈ R` with `r ^ n → 0`. -/
```

Comment:
This is a homemorphism, right? It currently seems painful to state this, but wouldn't this be useful when we move towards the weight space? In this case, it might simply be a ToDo.

Reply:
This is indeed a homeomorphism, for some natural choice of the topologies involved; but I'm not sure how useful a statement that is. I'd rather just get the definition of the map there for now, and worry about topologies later.

Reply:
Oh sure; I am not suggesting that you add it now, simply that you add a ToDo at the beginning of the file saying that this is something that could be done. I think it can be useful, and at any rate a quite natural result.